File: backend/app/Models/Product.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Str;
use Illuminate\Database\Eloquent\Builder;

class Product extends Model
{
    use SoftDeletes;

    protected $fillable = [
        'category_id',
        'name',
        'slug',
        'description',
        'price',
        'stock',
        'image_url',
        'image_public_id',
        'rating',
        'is_active',
    ];

    protected $casts = [
        'price'     => 'decimal:2',
        'rating'    => 'decimal:1',
        'stock'     => 'integer',
        'is_active' => 'boolean',
    ];

    protected $appends = ['stock_status'];

    protected static function boot(): void
    {
        parent::boot();

        static::creating(function ($product) {
            $product->slug = Str::slug($product->name);
        });
    }

    /*
    |--------------------------------------------------------------------------
    | Relationships
    |--------------------------------------------------------------------------
    */

    public function category(): BelongsTo
    {
        return $this->belongsTo(Category::class);
    }

    public function orderItems(): HasMany
    {
        return $this->hasMany(OrderItem::class);
    }

    /*
    |--------------------------------------------------------------------------
    | Accessors
    |--------------------------------------------------------------------------
    */

    public function getImageUrlAttribute($value): string
    {
        return $value ?: asset('images/placeholder.png');
    }

    public function getStockStatusAttribute(): string
    {
        if ($this->stock <= 0) {
            return 'out_of_stock';
        }

        if ($this->stock <= 5) {
            return 'low_stock';
        }

        return 'in_stock';
    }

    /*
    |--------------------------------------------------------------------------
    | Scopes
    |--------------------------------------------------------------------------
    */

    public function scopeActive(Builder $query): Builder
    {
        return $query->where('is_active', true);
    }

    public function scopeInCategory(Builder $query, ?string $categorySlug): Builder
    {
        if ($categorySlug && $categorySlug !== 'all') {
            return $query->whereHas('category', function ($q) use ($categorySlug) {
                $q->where('slug', $categorySlug);
            });
        }

        return $query;
    }

    public function scopeSearch(Builder $query, ?string $term): Builder
    {
        if ($term) {
            return $query->where(function ($q) use ($term) {
                $q->where('name', 'like', "%{$term}%")
                  ->orWhere('description', 'like', "%{$term}%");
            });
        }

        return $query;
    }
}

File: admin_web/lib/data/models/product.dart
import 'category.dart';

class Product {
  final int id;
  final int categoryId;
  final String name;
  final String slug;
  final String description;
  final double price;
  final int stock;
  final String imageUrl;
  final String imagePublicId;
  final double rating;
  final bool isActive;
  final String stockStatus;
  final Category category;
  final String priceFormatted;

  Product({
    required this.id,
    required this.categoryId,
    required this.name,
    required this.slug,
    required this.description,
    required this.price,
    required this.stock,
    required this.imageUrl,
    required this.imagePublicId,
    required this.rating,
    required this.isActive,
    required this.stockStatus,
    required this.category,
    required this.priceFormatted,
  });
factory Product.fromJson(Map<String, dynamic> json) {
  return Product(
    id: json['id'] ?? 0,
    categoryId: json['category_id'] ?? 0,
    name: json['name'] ?? '',
    slug: json['slug'] ?? '',
    description: json['description'] ?? '',
    price: double.tryParse(json['price']?.toString() ?? '') ?? 0.0,
    stock: json['stock'] ?? 0,
    imageUrl: json['image_url'] ?? '',
    imagePublicId: json['image_public_id'] ?? '',
    rating: double.tryParse(json['rating']?.toString() ?? '') ?? 0.0,
    isActive: json['is_active'] == true || json['is_active'] == 1,
    stockStatus: json['stock_status'] ?? '',
    category: json['category'] != null
        ? Category.fromJson(json['category'])
        : Category(
            id: 0,
            name: 'Unknown',
            slug: 'unknown',         
            color: '#CCCCCC',
            isActive: false,          
            productsCount: 0,         
          ),
    priceFormatted: json['price_formatted'] ?? '',
  );
}



  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'category_id': categoryId,
      'name': name,
      'slug': slug,
      'description': description,
      'price': price,
      'stock': stock,
      'image_url': imageUrl,
      'image_public_id': imagePublicId,
      'rating': rating,
      'is_active': isActive,
      'stock_status': stockStatus,
      'category': category.toJson(),
      'price_formatted': priceFormatted,
    };
  }
}
